### PR TITLE
Fix issues with apostrophe being passed in table name + Improvements in SQLServerParameterMetadata

### PR DESCRIPTION
**Pull request by cheenamalhotra**
- Fixes issue #775 
- Fixes Null Pointer Exception reported in #778
- Updates Junit tests to test apostrophe(') in randomized table name

### CODE REVIEW

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java [hunk 1]
```diff
@@ -2123,8 +2124,9 @@ public final void clearBatch() throws SQLServerException {
                 try (SQLServerStatement stmt = (SQLServerStatement) connection.createStatement(
                         ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY, connection.getHoldability(),
                         stmtColumnEncriptionSetting);
-                        SQLServerResultSet rs = stmt.executeQueryInternal(
-                                "sp_executesql N'SET FMTONLY ON SELECT * FROM " + tableName + " '");) {
+                        SQLServerResultSet rs = stmt
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java]: rs not closed

Comment (cheenamalhotra) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java]: Its inside try-w-resource try block - hence does get closed

Comment (lilgreenbird) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java]: oops missed that

File: src/main/java/com/microsoft/sqlserver/jdbc/Util.java [hunk 2]
```diff
@@ -985,6 +985,17 @@ static synchronized boolean checkIfNeedNewAccessToken(SQLServerConnection connec
...
     }
+
+    /**
+     * Escapes single quotes (') in object name to convert and pass it as String safely.
+     * 
+     * @param name
+     *        Object name to be passed as String
+     * @return Converted object name
+     */
+    static String escapeQuotes(String name) {
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/Util.java]: Can we rename this method to `escapeSingleQuotes`?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerParameterMetaData.java [hunk 3]
```diff
@@ -208,91 +206,98 @@ private String parseInsertColumns(String sql, String columnMarker) throws SQLSer
...
-                } else {
-                    qm.precision = rsQueryMeta.getInt("suggested_precision");
-                    qm.scale = rsQueryMeta.getInt("suggested_scale");
-
-                    Matcher matcher = datatypePattern.matcher(typename);
-                    if (matcher.matches()) {
-                        // the datatype has some precision/scale defined explicitly.
-                        ssType = SSType.of(matcher.group(1));
-                        if (typename.equalsIgnoreCase("varchar(max)") || typename.equalsIgnoreCase("varbinary(max)")) {
+            if (null != rsQueryMeta) {
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerParameterMetaData.java]: Don't we need the same check for `parseQueryMetaFor2008`?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerParameterMetaData.java [hunk 4]
```diff
@@ -557,28 +561,47 @@ private void checkClosed() throws SQLServerException {
...
-                }
-                rsProcedureMeta.beforeFirst();
-
-                // Sixth is DATA_TYPE
-                rsProcedureMeta.getColumn(6).setFilter(new DataTypeFilter());
-                if (con.isKatmaiOrLater()) {
-                    rsProcedureMeta.getColumn(8).setFilter(new ZeroFixupFilter());
-                    rsProcedureMeta.getColumn(9).setFilter(new ZeroFixupFilter());
-                    rsProcedureMeta.getColumn(17).setFilter(new ZeroFixupFilter());
+                // DO NOT close the Statement as resultSet 'rsProcedureMeta' is required by other APIs
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerParameterMetaData.java]: Try-with-resources will close the statement though. Is this comment still accurate?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerParameterMetaData.java [hunk 5]
```diff
@@ -685,25 +704,26 @@ public boolean isWrapperFor(Class<?> iface) throws SQLException {
...
 
-    private void verifyParameterPosition(int param) throws SQLServerException {
+    private Map<String, Object> getParameterInfo(int param) throws SQLServerException {
         boolean bFound = false;
-        try {
-            if (((SQLServerPreparedStatement) stmtParent).bReturnValueSyntax && isTVP) {
-                bFound = rsProcedureMeta.absolute(param);
-            } else {
-                bFound = rsProcedureMeta.absolute(param + 1); // Note row 1 is the 'return value' meta data
+        if ((stmtParent).bReturnValueSyntax && isTVP) {
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerParameterMetaData.java]: I can't figure out why the variable would be named `bFound`. Please rename it to a descriptive one.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerParameterMetaData.java [hunk 6]
```diff
@@ -557,28 +561,47 @@ private void checkClosed() throws SQLServerException {
...
+                        rsProcedureMeta.getColumn(17).setFilter(new ZeroFixupFilter());
+                    }
+
+                    procMetadata = new ArrayList<>();
+
+                    // Process ResultSet Procedure Metadata for API usage
+                    while (rsProcedureMeta.next()) {
+                        procMetadata.add(new HashMap<String, Object>() {
+                            {
+                                put("DATA_TYPE", rsProcedureMeta.getShort("DATA_TYPE"));
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerParameterMetaData.java]: I think we should have some kind of an exception handling here. 
If one of the `ResultSet` getters fails here, the exception message will be misleading. The `catch` block you removed below can be moved up here.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerParameterMetaData.java [hunk 7]
```diff
@@ -735,34 +754,25 @@ public String getParameterClassName(int param) throws SQLServerException {
...
-            }
-        } catch (SQLException e) {
-            SQLServerException.makeFromDriverError(con, stmtParent, e.toString(), null, false);
-            return 0;
+        if (procMetadata == null) {
+            // PreparedStatement
+            return queryMetaMap.size();
+        } else {
+            // Row 1 is Return Type metadata
+            return procMetadata.size() - 1;
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerParameterMetaData.java]: what if the size is 0?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerParameterMetaData.java (hunk 6 above)
Comment (cheenamalhotra) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerParameterMetaData.java]: There is a catch block in this method at the end. The exception handling that was done for these getters is done same way here.

Comment (ulvii) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerParameterMetaData.java]: Not really, the catch block will just re-throw the exception. 
```
catch (SQLServerException e) {
    throw e;
}
```

Comment (cheenamalhotra) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerParameterMetaData.java]: So what's the problem with that? We don't need to wrap SQLServerException again!

**Pull request merged**